File: app/src/main/java/com/sumian/sd/network/response/ErrorResponse.java
package com.sumian.sd.network.response;

import com.google.gson.annotations.SerializedName;

/**
 * Created by jzz
 * on 2017/9/30
 * <p>
 * desc:
 */

public class ErrorResponse {
    @SerializedName("status_code")
    private int code;
    private String message;
    // 兼容性数据，有的网络请求有error info，有个没有
    private ErrorInfo error;

    public int getCode() {
        if (error != null) {
            return error.getCode();
        }
        return code;
    }

    public void setCode(int code) {
        this.code = code;
    }

    public String getMessage() {
        if (error != null) {
            return error.getUser_message();
        }
        return message;
    }

    public void setMessage(String message) {
        this.message = message;
    }

    public ErrorInfo getError() {
        return error;
    }

    public void setError(ErrorInfo error) {
        this.error = error;
    }

    @Override
    public String toString() {
        return "ErrorResponse{" +
                "code=" + code +
                ", message='" + message + '\'' +
                '}';
    }
}

File: app/src/main/java/com/sumian/sd/network/response/ErrorResponse.kt
package com.sumian.sd.network.response

import com.google.gson.annotations.SerializedName

/**
 * <pre>
 *     @author : Zhan Xuzhao
 *     e-mail : xuzhao.z@sumian.com
 *     time   : 2018/6/14 10:29
 *     desc   :
 *     version: 1.0
 * </pre>
 */
data class ErrorResponse(var code: Int, var message: String) {
    companion object {

        const val STATUS_CODE_ERROR_UNKNOWN = 0
        const val STATUS_CODE_NOT_FOUND = 404

        fun createFromErrorInfo(errorInfo: ErrorInfo499?): ErrorResponse? {
            return errorInfo?.let { ErrorResponse(errorInfo.error.code, errorInfo.error.userMessage) }
        }

        fun createFromErrorInfo(errorInfo: ErrorInfo400?): ErrorResponse? {
            return errorInfo?.let { ErrorResponse(errorInfo.statusCode, errorInfo.message) }
        }
    }

    fun isNotFound(): Boolean {
        return code == STATUS_CODE_NOT_FOUND
    }
}

data class ErrorInfo499(
        @SerializedName("error") val error: Error
)

data class Error(
        @SerializedName("code") val code: Int,
        @SerializedName("user_message") val userMessage: String,
        @SerializedName("internal_message") val internalMessage: String,
        @SerializedName("more_info") val moreInfo: Any
)

data class ErrorInfo400(
        @SerializedName("message") val message: String,
        @SerializedName("status_code") val statusCode: Int
)